File: gin/services/test_service.go
package services

import (
	"archive/zip"
	"encoding/csv"
	"errors"
	"experiment/models"
	"experiment/repositories"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"time"
)

type TestService struct {
	//Repo 				   repositories.
	UserRepository           *repositories.UserRepository
	GroupRepository          *repositories.GroupRepository
	UsergroupstateRepository *repositories.UsergroupstateRepository
	TempFiles                map[string]string
}

//func NewTestService(repo *repositories.repositories) *TestService {
//	return &TestService{
//		Repo:      repo,
//		TempFiles: make(map[string]string),
//	}
//}

func (fs *TestService) HandleFileUpload(filePath string) error {
	err := fs.OpenZipFile(filePath)
	if err != nil {
		return err
	}
	return fs.ImportFromFile()
}

func (fs *TestService) OpenZipFile(file string) error {
	fs.TempFiles = make(map[string]string)
	r, err := zip.OpenReader(file)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, file := range r.File {
		if filepath.Ext(file.Name) == ".csv" {
			outFile, err := os.CreateTemp("", file.Name)
			if err != nil {
				return err
			}
			fs.TempFiles[file.Name] = outFile.Name()
			src, _ := file.Open()
			defer src.Close()
			_, err = io.Copy(outFile, src)
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func (fs *TestService) ImportFromFile() error {
	err := fs.insertGroupState()
	if err != nil {
		return err
	}
	err = fs.insertGroup()
	if err != nil {
		return err
	}
	return fs.insertUser()
}

func (fs *TestService) insertGroupState() error {
	filePath, exists := fs.TempFiles["usergroupstate.csv"]
	if !exists {
		return errors.New("usergroupstate not found")
	}
	file, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	_, _ = reader.Read()
	for {
		line, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
		id, _ := strconv.Atoi(line[0])
		state := models.Usergroupstate{ID: id, Label: line[1]}
		fs.UsergroupstateRepository.Save(state)
	}
	return nil
}

func (fs *TestService) insertGroup() error {
	filePath, exists := fs.TempFiles["group.csv"]
	if !exists {
		return errors.New("group not found")
	}
	file, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	_, err = reader.Read() // skip header
	if err != nil {
		return err
	}

	for {
		line, err := reader.Read()
		if err != nil {
			break
		}

		groupID, _ := strconv.Atoi(line[0])
		contractSigned, _ := time.Parse("2006-01-02 15:04:05.000", line[2])
		contractNumber, _ := strconv.Atoi(line[3])
		priority, _ := strconv.ParseFloat(line[16], 64)

		stateID, _ := strconv.Atoi(line[14])
		state, err := fs.UsergroupstateRepository.FindById(stateID)
		if err != nil {
			return err
		}

		group := models.Group{
			ID:                     groupID,
			Name:                   line[1],
			ContractSigned:         contractSigned,
			ContractNumber:         &contractNumber,
			InstituteName:          &line[4],
			InstituteAddressName:   line[5],
			InstituteAddressNumber: &line[6],
			InstitutePostalCode:    line[7],
			Province:               &line[8],
			City:                   line[9],
			Country:                line[10],
			VAT:                    &line[11],
			InstitutePhone:         line[12],
			InstituteEmail:         line[13],
			State:                  state,
			Priority:               priority,
		}

		err = fs.GroupRepository.Save(group)
		if err != nil {
			return err
		}
	}
	return nil
}

func (fs *TestService) insertUser() error {
	filePath, exists := fs.TempFiles["user.csv"]
	if !exists {
		return errors.New("user not found")
	}
	file, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer file.Close()
	reader := csv.NewReader(file)
	_, err = reader.Read() // skip header
	if err != nil {
		return err
	}

	for {
		line, err := reader.Read()
		if err != nil {
			break
		}

		userID, _ := strconv.Atoi(line[0])
		groupID, _ := strconv.Atoi(line[1])
		stateID, _ := strconv.Atoi(line[2])
		isDsmAdmin, _ := strconv.ParseBool(line[9])
		isGroupAdmin, _ := strconv.ParseBool(line[10])
		isPlanner, _ := strconv.ParseBool(line[11])
		isSupportTech, _ := strconv.ParseBool(line[12])

		group, err := fs.GroupRepository.FindById(groupID)
		if err != nil {
			return err
		}

		state, err := fs.UsergroupstateRepository.FindById(stateID)
		if err != nil {
			return err
		}

		user := models.User{
			ID:            userID,
			Group:         group,
			State:         state,
			Name:          line[3],
			Surname:       line[4],
			Login:         line[5],
			Passwd:        &line[6],
			Email:         line[7],
			Phone:         &line[8],
			IsDsmAdmin:    isDsmAdmin,
			IsGroupAdmin:  isGroupAdmin,
			IsPlanner:     isPlanner,
			IsSupportTech: isSupportTech,
			Photo:         &line[13],
		}

		err = fs.UserRepository.Save(user)
		if err != nil {
			return err
		}
	}
	return nil
}

func openFile(fs *TestService, tableName string) (*os.File, error) {

	return nil, nil
}

File: spring-boot/src/main/java/com/bp/experiment/service/TestService.java
package com.bp.experiment.service;

import java.io.BufferedOutputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import com.bp.experiment.dao.model.Group;
import com.bp.experiment.dao.model.User;
import com.bp.experiment.dao.model.Usergroupstate;
import com.bp.experiment.dao.repository.GroupRepository;
import com.bp.experiment.dao.repository.UserRepository;
import com.bp.experiment.dao.repository.UsergroupstateRepository;
import com.opencsv.CSVParserBuilder;
import com.opencsv.CSVReader;
import com.opencsv.CSVReaderBuilder;
import com.opencsv.exceptions.CsvValidationException;
import jakarta.persistence.EntityManager;
import jakarta.transaction.Transactional;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class TestService {


    private final EntityManager entityManager;
    private final UsergroupstateRepository usergroupstateRepository;
    private final GroupRepository groupRepository;
    private final UserRepository userRepository;

    @Getter
    private final Map<String, Path> tempFilesByOriginalFileName = new HashMap<>();

    public void handleFileUpload(MultipartFile file) throws CsvValidationException, IOException, ParseException {
        openZipFile(file);
        importFromFile();
    }

    @Transactional
    public void importFromFile() throws CsvValidationException, IOException, ParseException {
        insertGroupState();
        insertGroup();
        insertUser();
    }

    private void insertGroupState() throws CsvValidationException, IOException {
        try {
            InputStream file = new FileInputStream(getTempFile("usergroupstate.csv").toFile());
            CSVReader reader = buildRuianCsvReader(new InputStreamReader(file));
            // skip head
            String[] line = reader.readNext();
            while ((line = reader.readNext()) != null) {
                Usergroupstate usergroupstate = Usergroupstate.builder().id(Integer.parseInt(line[0])).label(line[1]).build();
                usergroupstateRepository.save(usergroupstate);
                entityManager.clear();
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    private void insertGroup() throws CsvValidationException, IOException, ParseException {
        try {
            InputStream file = new FileInputStream(getTempFile("group.csv").toFile());
            CSVReader reader = buildRuianCsvReader(new InputStreamReader(file));
            // skip head
            String[] line = reader.readNext();
            while ((line = reader.readNext()) != null) {
                Group group = Group.builder().id(Integer.parseInt(line[0])).name(line[1]).contractSigned(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").parse(line[2]).toInstant()).contractNumber(Integer.parseInt(line[3])).instituteName(line[4]).instituteAddressName(line[5]).instituteAddressNumber(line[6]).institutePostalCode(line[7]).province(line[8]).city(line[9]).country(line[10]).vat(line[11]).institutePhone(line[12]).instituteEmail(line[13]).state(usergroupstateRepository.findById(Integer.parseInt(line[14])).get()).priority(Float.parseFloat(line[16])).build();
                groupRepository.save(group);
                entityManager.clear();
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    private void insertUser() {
        try {
            InputStream file = new FileInputStream(getTempFile("user.csv").toFile());
            CSVReader reader = buildRuianCsvReader(new InputStreamReader(file));
            // skip head
            String[] line = reader.readNext();
            while ((line = reader.readNext()) != null && line.length > 0) {
                User user = User.builder().id(Integer.parseInt(line[0])).group(groupRepository.findById(Integer.parseInt(line[1])).get()).state(usergroupstateRepository.findById(Integer.parseInt(line[2])).get()) // Assuming Usergroupstate has a constructor with id
                        .name(line[3]).surname(line[4]).login(line[5]).passwd(line[6]).email(line[7]).phone(line[8]).isDsmAdmin(Boolean.parseBoolean(line[9])).isGroupAdmin(Boolean.parseBoolean(line[10])).isPlanner(Boolean.parseBoolean(line[11])).isSupportTech(Boolean.parseBoolean(line[12])).photo(line[13]).build();
                userRepository.save(user);
                entityManager.clear();
            }
        } catch (IOException | CsvValidationException e) {
            throw new RuntimeException(e);
        }
    }

    private CSVReader buildRuianCsvReader(Reader reader) {
        return new CSVReaderBuilder(reader).withCSVParser(new CSVParserBuilder().withSeparator(',').build()).build();
    }

    private void openZipFile(MultipartFile file) {
        try (ZipInputStream zis = new ZipInputStream(file.getInputStream())) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                if (!entry.isDirectory() && entry.getName().endsWith(".csv")) {
                    Path tempFile = createTempFile(entry.getName());
                    try (BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(tempFile.toFile()))) {
                        byte[] buffer = new byte[1024];
                        int len;
                        while ((len = zis.read(buffer)) > 0) {
                            bos.write(buffer, 0, len);
                        }
                    }
                }
                zis.closeEntry();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private Path createTempFile(String fileName) {
        try {
            Path tempFile = Files.createTempFile("tmp", fileName);
            tempFilesByOriginalFileName.put(fileName, tempFile);
            return tempFile;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private Path getTempFile(String filePath) {
        return tempFilesByOriginalFileName.get(filePath);
    }

}
